### PR TITLE
fix: Add compatibility with protobuf==5.x

### LINKED ISSUE #425: Presubmits fail with `RecursionError: maximum recursion depth exceeded while calling a Python object` when using protobuf at HEAD
See PR https://github.com/googleapis/proto-plus-python/pull/423
See build log here: https://github.com/googleapis/proto-plus-python/actions/runs/7576004395/job/20633856887?pr=423

```
____________________________________________________________________________________________________ test_message_pb ____________________________________________________________________________________________________

    def test_message_pb():
>       class Foo(proto.Message):

tests/test_message.py:205: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
proto/message.py:275: in __new__
    file_info.generate_file_pb(new_class=cls, fallback_salt=full_name)
proto/_file_info.py:113: in generate_file_pb
    pb_message = reflection.GeneratedProtocolMessageType(
.nox/unit-3-10/lib/python3.10/site-packages/google/protobuf/internal/python_message.py:166: in __init__
    descriptor.GetOptions().message_set_wire_format):
.nox/unit-3-10/lib/python3.10/site-packages/google/protobuf/descriptor.py:239: in GetOptions
    self._LazyLoadOptions()
.nox/unit-3-10/lib/python3.10/site-packages/google/protobuf/descriptor.py:219: in _LazyLoadOptions
    features = self._ResolveFeatures(
.nox/unit-3-10/lib/python3.10/site-packages/google/protobuf/descriptor.py:176: in _ResolveFeatures
    features.CopyFrom(self._parent._GetFeatures())
.nox/unit-3-10/lib/python3.10/site-packages/google/protobuf/descriptor.py:158: in _GetFeatures
    self._LazyLoadOptions()
.nox/unit-3-10/lib/python3.10/site-packages/google/protobuf/descriptor.py:205: in _LazyLoadOptions
    features = self._ResolveFeatures(
.nox/unit-3-10/lib/python3.10/site-packages/google/protobuf/descriptor.py:180: in _ResolveFeatures
    unresolved.CopyFrom(raw_options.features)
.nox/unit-3-10/lib/python3.10/site-packages/google/protobuf/internal/python_message.py:744: in getter
    field_value = field._default_constructor(self)
.nox/unit-3-10/lib/python3.10/site-packages/google/protobuf/internal/python_message.py:450: in MakeSubMessageDefault
    message_factory.GetMessageClass(message_type)
.nox/unit-3-10/lib/python3.10/site-packages/google/protobuf/message_factory.py:50: in GetMessageClass
    return _InternalCreateMessageClass(descriptor)
.nox/unit-3-10/lib/python3.10/site-packages/google/protobuf/message_factory.py:109: in _InternalCreateMessageClass
    result_class = _GENERATED_PROTOCOL_MESSAGE_TYPE(
.nox/unit-3-10/lib/python3.10/site-packages/google/protobuf/internal/python_message.py:174: in __init__
    _AttachFieldHelpers(cls, field)
.nox/unit-3-10/lib/python3.10/site-packages/google/protobuf/internal/python_message.py:269: in _AttachFieldHelpers
    is_repeated = field_descriptor.label == _FieldDescriptor.LABEL_REPEATED
.nox/unit-3-10/lib/python3.10/site-packages/google/protobuf/descriptor.py:722: in label
    self._GetFeatures().field_presence
.nox/unit-3-10/lib/python3.10/site-packages/google/protobuf/descriptor.py:158: in _GetFeatures
    self._LazyLoadOptions()
.nox/unit-3-10/lib/python3.10/site-packages/google/protobuf/descriptor.py:219: in _LazyLoadOptions
    features = self._ResolveFeatures(
.nox/unit-3-10/lib/python3.10/site-packages/google/protobuf/descriptor.py:176: in _ResolveFeatures
    features.CopyFrom(self._parent._GetFeatures())
.nox/unit-3-10/lib/python3.10/site-packages/google/protobuf/descriptor.py:158: in _GetFeatures
    self._LazyLoadOptions()
.nox/unit-3-10/lib/python3.10/site-packages/google/protobuf/descriptor.py:205: in _LazyLoadOptions
    features = self._ResolveFeatures(
.nox/unit-3-10/lib/python3.10/site-packages/google/protobuf/descriptor.py:180: in _ResolveFeatures
    unresolved.CopyFrom(raw_options.features)
.nox/unit-3-10/lib/python3.10/site-packages/google/protobuf/internal/python_message.py:744: in getter
    field_value = field._default_constructor(self)
.nox/unit-3-10/lib/python3.10/site-packages/google/protobuf/internal/python_message.py:450: in MakeSubMessageDefault
    message_factory.GetMessageClass(message_type)
.nox/unit-3-10/lib/python3.10/site-packages/google/protobuf/message_factory.py:50: in GetMessageClass
    return _InternalCreateMessageClass(descriptor)
E   RecursionError: maximum recursion depth exceeded while calling a Python object
!!! Recursion detected (same locals & position)
```

### LINKED ISSUE #427: Add support for pre-release testing of dependencies
There is a `prerelease_deps` nox session for packages in `google-cloud-python`, [for example](https://github.com/googleapis/google-cloud-python/blob/9f0608987b8b72f9e613b170b91b6ab40493287f/packages/google-cloud-access-approval/noxfile.py#L340), that tests against pre-release versions of direct dependencies. We should create a similar presubmit check in this repository that tests pre-release versions of the direct dependencies shown below.

https://github.com/googleapis/proto-plus-python/blob/325f322cdc1f16777028d889f2c92e4a54dcde2b/setup.py#L45-L50

For example, when a new pre-release version of `google-api-core` or `protobuf` is published, there should be a prerelease presubmit check in this repo that run tests against it.

### LINKED ISSUE #431: TypeError: MessageToJson() got an unexpected keyword argument 'including_default_value_fields'
The following stack trace below appears when testing `protobuf==5.26.0rc2` : https://pypi.org/project/protobuf/5.26.0rc2/

The error is occurring because the `including_default_value_fields` argument was renamed in `MessageToJson` and `MessageToDict`.

Related commit in upstream protobuf repository: https://github.com/protocolbuffers/protobuf/commit/26995798757fbfef5cf6648610848e389db1fecf

```
    def to_json(
        cls,
        instance,
        *,
        use_integers_for_enums=True,
        including_default_value_fields=True,
        preserving_proto_field_name=False,
        sort_keys=False,
        indent=2,
        float_precision=None,
    ) -> str:
        """Given a message instance, serialize it to json
    
        Args:
            instance: An instance of this message type, or something
                compatible (accepted by the type's constructor).
            use_integers_for_enums (Optional(bool)): An option that determines whether enum
                values should be represented by strings (False) or integers (True).
                Default is True.
            preserving_proto_field_name (Optional(bool)): An option that
                determines whether field name representations preserve
                proto case (snake_case) or use lowerCamelCase. Default is False.
            sort_keys (Optional(bool)): If True, then the output will be sorted by field names.
                Default is False.
            indent (Optional(int)): The JSON object will be pretty-printed with this indent level.
                An indent level of 0 or negative will only insert newlines.
                Pass None for the most compact representation without newlines.
            float_precision (Optional(int)): If set, use this to specify float field valid digits.
                Default is None.
        Returns:
            str: The json string representation of the protocol buffer.
        """
>       return MessageToJson(
            cls.pb(instance),
            use_integers_for_enums=use_integers_for_enums,
            including_default_value_fields=including_default_value_fields,
            preserving_proto_field_name=preserving_proto_field_name,
            sort_keys=sort_keys,
            indent=indent,
            float_precision=float_precision,
        )
E       TypeError: MessageToJson() got an unexpected keyword argument 'including_default_value_fields'
```

### LINKED ISSUE #445: Add support for protobuf 5.x
Creating a tracking issue to add support for protobuf 5.x which was released last month

https://pypi.org/project/protobuf/5.26.0/

### PR DESCRIPTION
**Pull request by parthea**
BEGIN_COMMIT_OVERRIDE
feat: Add `always_print_fields_with_no_presence` fields to `to_json` and `to_dict`
docs: Deprecate field `including_default_value_fields` in `to_json` and `to_dict`
fix: Add compatibility with protobuf==5.x
fix: AttributeError module 'google._upb._message' has no attribute 'MessageMapContainer'
ci: Add support for pre-release sessions
ci: Add python 3.13 to testing matrix
ci: Resolve `RecursionError: maximum recursion depth exceeded while calling a Python object` when running tests
END_COMMIT_OVERRIDE

Fixes https://github.com/googleapis/proto-plus-python/issues/425
Fixes https://github.com/googleapis/proto-plus-python/issues/427
Fixes https://github.com/googleapis/proto-plus-python/issues/431
Fixes https://github.com/googleapis/proto-plus-python/issues/445

For compatibility with https://pypi.org/project/protobuf/5.26.0rc2/

See https://github.com/protocolbuffers/protobuf/commit/26995798757fbfef5cf6648610848e389db1fecf

### CODE REVIEW

File: proto/message.py [hunk 1]
```diff
@@ -388,6 +392,11 @@ def to_json(
             use_integers_for_enums (Optional(bool)): An option that determines whether enum
                 values should be represented by strings (False) or integers (True).
                 Default is True.
+            including_default_value_fields (Optional(bool)): Deprecated. Use argument
+                `always_print_fields_with_no_presence` instead. An option that
+                determines whether the default field values should be included in the results.
+                Default is True. If `always_print_fields_with_no_presence` is set to False, this
```
Comment (vchudnov-g) [inline proto/message.py:398]: Both of these fields say that they don't take effect if the other is False. This is confusing enough that it's not obvious what the right semantics are and whether they make sense.

I recommend setting the default value for both fields in the arg list above to be `None`, and then using them thus:

```
if ((always_print_fields_with_no_presence is not None) and 
    (including_default_value_fields is not None) and
    (always_print_fields_with_no_presence != including_default_value_fields):
   raise ValueError("new and deprecated parameters are inconsistent")

print_field = (always_print_fields_with_no_presence if always_print_fields_with_no_presence is not None
               else (including_default_value_fields if including_default_value_fields is not None
                     else True)

```

and simplifying the comments for both parameters to say that any value explicitly provided for one must be consistent with the any value explicitly provided for the other.

File: proto/message.py [hunk 2]
```diff
@@ -398,18 +407,43 @@ def to_json(
...
-            indent=indent,
-            float_precision=float_precision,
-        )
+        # For backwards compatibility of this breaking change in protobuf 5.x which is specific to proto2
+        # https://github.com/protocolbuffers/protobuf/commit/26995798757fbfef5cf6648610848e389db1fecf
+        if PROTOBUF_VERSION[0] in ("3", "4"):
+            return MessageToJson(
+                cls.pb(instance),
+                use_integers_for_enums=use_integers_for_enums,
+                including_default_value_fields=including_default_value_fields,
```
Comment (vchudnov-g) [inline proto/message.py:424]: If we're deprecating this field on thisA PI surface, we should have the same deprecation logic on the RHS here as in the `else` branch (and I suggest the `print_field` variable I defined in my other comment).

File: proto/message.py [hunk 3]
```diff
@@ -448,24 +483,49 @@ def to_dict(
             preserving_proto_field_name (Optional(bool)): An option that
                 determines whether field name representations preserve
                 proto case (snake_case) or use lowerCamelCase. Default is True.
-            including_default_value_fields (Optional(bool)): An option that
+            including_default_value_fields (Optional(bool)): Deprecated. Use argument
```
Comment (vchudnov-g) [inline proto/message.py:486]: same comments as in `to_json`

**Comment by parthea** [comment]
The pre-release failure is captured in https://github.com/googleapis/proto-plus-python/issues/444

File: proto/message.py (hunk 3 above)
Comment (parthea) [reply proto/message.py:495]: Fixed in https://github.com/googleapis/proto-plus-python/pull/433/commits/0e65d18eefc21ef7b34b9fc12f9aa779affca437

File: proto/message.py (hunk 2 above)
Comment (parthea) [reply proto/message.py]: Fixed in https://github.com/googleapis/proto-plus-python/pull/433/commits/0e65d18eefc21ef7b34b9fc12f9aa779affca437

File: tests/test_fields_mitigate_collision.py [hunk 4]
```diff
@@ -15,6 +15,7 @@
 import proto
 import pytest
 
+
```
Comment (parthea) [inline tests/test_fields_mitigate_collision.py:18]:
```suggestion
```

**Review by vchudnov-g — COMMENTED** [review]
Minor drive-by comments.

File: proto/message.py [hunk 5]
```diff
@@ -369,16 +373,95 @@ def deserialize(cls, payload: bytes) -> "Message":
...
+        """
+        if (
+            always_print_fields_with_no_presence is None
+            and including_default_value_fields is None
+        ):
+            print_fields = True
+        else:
+            print_fields = (
+                always_print_fields_with_no_presence or including_default_value_fields
+            )
```
[the remark this answers is not in the document]
Comment (parthea) [reply proto/message.py:451]: The suggestion didn't work because we need to use a default of `print_fields=True` if both `always_print_fields_with_no_presence` and `including_default_value_fields` are `None` (not set).  

We need to distinguish between not set and `False`, especially since we are raising an exception if `always_print_fields_with_no_presence` and `including_default_value_fieldsincluding_default_value_fields` are set and don't match.

File: proto/message.py [hunk 6]
```diff
@@ -398,19 +486,51 @@ def to_json(
...
-            indent=indent,
-            float_precision=float_precision,
+
+        cls._warn_if_including_default_value_fields_is_used_protobuf_5(
+            including_default_value_fields
+        )
+        cls._raise_if_print_fields_values_are_set_and_differ(
+            always_print_fields_with_no_presence, including_default_value_fields
+        )
+        print_fields = cls._normalize_print_fields_without_presense(
```
[the remark this answers is not in the document]
Comment (parthea) [reply proto/message.py:506]: Fixed in https://github.com/googleapis/proto-plus-python/pull/433/commits/d159837e86642dc8a11f97d15a6b301c1d6f0a36

File: proto/marshal/compat.py [hunk 7]
```diff
@@ -37,14 +40,24 @@
 repeated_scalar_types = (containers.RepeatedScalarFieldContainer,)
 map_composite_types = (containers.MessageMap,)
 
+# See https://github.com/protocolbuffers/protobuf/issues/16596
+# In protobuf 5, we will use the name of the class instead `MessageMapContainer`
+# See `map_composite_types_str`
+map_composite_types_str = ("MessageMapContainer",)
```
[the remark this answers is not in the document]
Comment (parthea) [reply proto/marshal/compat.py]: Fixed in https://github.com/googleapis/proto-plus-python/pull/433/commits/0121178ee405d4c555bebdf84c0294e30c809c78

File: proto/marshal/compat.py [hunk 8]
```diff
@@ -37,14 +40,24 @@
...
+
 if _message:
     repeated_composite_types += (_message.RepeatedCompositeContainer,)
     repeated_scalar_types += (_message.RepeatedScalarContainer,)
-    map_composite_types += (_message.MessageMapContainer,)
 
+    # See https://github.com/protocolbuffers/protobuf/issues/16596
+    # In protobuf 5, we will use the name of the class instead `MessageMapContainer`
```
[the remark this answers is not in the document]
Comment (parthea) [reply proto/marshal/compat.py]: Fixed in https://github.com/googleapis/proto-plus-python/pull/433/commits/da7db5b0c344581913d7a380943c6f5d0b2e00b4

File: proto/message.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (vchudnov-g) [reply proto/message.py:458]: Ah, gotcha. I missed that part.  I think the following would do it, though it's arguably less clear than what you have.

```suggestion
        print_fields = (
            always_print_fields_with_no_presence is None
            and including_default_value_fields is None
        ) or (always_print_fields_with_no_presence or 
              including_default_value_fields)
```

[the remark this answers is not in the document]
Comment (vchudnov-g) [reply proto/message.py:458]: (It's always weird to have 
```python
if x:
  y = true
```

We could also use a pattern a bit more explicit than my suggestion above
```python
y=x
x=x or (...)
```

[the remark this answers is not in the document]
Comment (parthea) [reply proto/message.py]: Fixed in https://github.com/googleapis/proto-plus-python/pull/433/commits/4e9f90e14104987984f0825fbba4a43dae8dff9b

**Comment by parthea** [comment]
@vchudnov-g @ohmayr PTAL

File: tests/test_json.py [hunk 9]
```diff
@@ -121,6 +121,57 @@ class Squid(proto.Message):
...
+        Squid.to_json(
+            s,
+            including_default_value_fields=True,
+            always_print_fields_with_no_presence=True,
+        )
+        .replace(" ", "")
+        .replace("\n", "")
+    )
+    assert (
+        json2 == '{"name":"Steve","massKg":0}' or json2 == '{"massKg":0,"name":"Steve"}'
```
Comment (vchudnov-g) [inline tests/test_json.py:172]: Aren't dicts equal regardless of key order?

File: tests/test_message.py [hunk 10]
```diff
@@ -256,6 +256,13 @@ class Color(proto.Enum):
     s_dict = Squid.to_dict(s)
     assert s_dict["chromatophores"][0]["color"] == 1
 
+    s_dict = Squid.to_dict(
+        s,
+        including_default_value_fields=True,
+        always_print_fields_with_no_presence=True,
+    )
+    assert s_dict["chromatophores"][0]["color"] == 1
```
Comment (vchudnov-g) [inline tests/test_message.py:264]: What is this testing? I think we want to check that a non-optional field with the default value is emitted in the dictionary. So we should check assert that  a color value of `UNKNOWN` shows up under these conditions

And when the parameter is false, we should assert that a color value of `UNKNOWN` does NOT show.

I suggest for both of these cases, we modify the class to have two scalar fields, one of which we don't set (or set to the default), and test the four combinations that way. Essentially what we have in the test for `to_json`

File: tests/test_json.py [hunk 11]
```diff
@@ -121,6 +121,57 @@ class Squid(proto.Message):
...
+        Squid.to_json(s, always_print_fields_with_no_presence=False)
+        .replace(" ", "")
+        .replace("\n", "")
+    )
+    assert json1 == '{"name":"Steve"}'
+
+    json1 = (
+        Squid.to_json(
+            s,
+            including_default_value_fields=False,
```
Comment (vchudnov-g) [inline tests/test_json.py:134]: I would add two additional cases: when only `including_default_value_fields` is set to `False`, and when only it is set to `True` (i.e. it is the only parameter specified).

(and ditto in test for `to_dict`)

File: tests/test_json.py [hunk 12]
```diff
@@ -121,6 +121,57 @@ class Squid(proto.Message):
...
+            s,
+            including_default_value_fields=True,
+            always_print_fields_with_no_presence=True,
+        )
+        .replace(" ", "")
+        .replace("\n", "")
+    )
+    assert (
+        json2 == '{"name":"Steve","massKg":0}' or json2 == '{"massKg":0,"name":"Steve"}'
+    )
```
Comment (vchudnov-g) [inline tests/test_json.py:173]: No action needed--just a comment. I was going to suggest we add test cases to ensure an empty `repeated` is printed only when the two parameters we're testing are true, but that gets into testing protobuf functionality. I think the other test cases you have and I suggest are enough to show we're presenting the right parameters to protobuf.

File: proto/marshal/compat.py [hunk 13]
```diff
@@ -19,6 +19,9 @@
 # not be included.
 
 from google.protobuf.internal import containers
+import google.protobuf
+
+PROTOBUF_VERSION = google.protobuf.__version__
```
Comment (vchudnov-g) [inline proto/marshal/compat.py:24]: Not a blocker, but just a reminder that we could have a couple of lines here to make the version into a tuple of mostly numbers, as we've talked about in a different context. That would get rid of string comparisons and list membership checks below.

Not a biggie; we can defer until we do this consistently across the various repos.

File: noxfile.py [hunk 14]
```diff
@@ -71,17 +72,59 @@ def unit(session, proto="python"):
...
 
 
-# Check if protobuf has released wheels for new python versions
-# https://pypi.org/project/protobuf/#files
-# This list will generally be shorter than 'unit'
-@nox.session(python=["3.6", "3.7", "3.8", "3.9", "3.10"])
-def unitcpp(session):
-    return unit(session, proto="cpp")
+# Only test upb and python implementation backends.
+# As of protobuf 4.x, the "ccp" implementation is not available in the PyPI pacakge as per
```
[the remark this answers is not in the document]
Comment (parthea) [reply noxfile.py]: Fixed in https://github.com/googleapis/proto-plus-python/pull/433/commits/71554fe37a1a26133d4f589130ed358315e9998a

File: noxfile.py [hunk 15]
```diff
@@ -30,33 +29,35 @@
...
     constraints_path = str(
         CURRENT_DIRECTORY / "testing" / f"constraints-{session.python}.txt"
     )
 
-    session.env["PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION"] = proto
+    session.env["PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION"] = implementation
     session.install("coverage", "pytest", "pytest-cov", "pytz")
     session.install("-e", ".[testing]", "-c", constraints_path)
-    if proto == "cpp":  # 4.20 does not have cpp.
+    if implementation == "cpp":  # 4.20 does not have cpp.
```
[the remark this answers is not in the document]
Comment (parthea) [reply noxfile.py]: Fixed in https://github.com/googleapis/proto-plus-python/pull/433/commits/7541099007ae6fbdee6cf8e067734fa9b4b414a3

[the remark this answers is not in the document]
Comment (parthea) [reply noxfile.py]: This appears to be an obsolete comment. I couldn't locate it in the branch.

[the remark this answers is not in the document]
Comment (parthea) [reply noxfile.py]: Essentially for `cpp` we need to downgrade the protobuf version to `protobuf < 4`. When the condition is false, we will install the latest protobuf version or whatever is defined in `constraints_path`

File: proto/message.py [hunk 16]
```diff
@@ -369,16 +373,101 @@ def deserialize(cls, payload: bytes) -> "Message":
...
+        if (
+            always_print_fields_with_no_presence is not None
+            and including_default_value_fields is not None
+            and always_print_fields_with_no_presence != including_default_value_fields
+        ):
+            raise ValueError(
+                "Arguments `always_print_fields_with_no_presence` and `including_default_value_fields` must match"
+            )
+
+    def _normalize_print_fields_without_presense(
```
[the remark this answers is not in the document]
Comment (parthea) [reply proto/message.py]: Fixed in https://github.com/googleapis/proto-plus-python/pull/433/commits/2da20df33ac5621f72769345f7187337c8598745

File: proto/message.py [hunk 17]
```diff
@@ -369,16 +373,101 @@ def deserialize(cls, payload: bytes) -> "Message":
...
+        cls,
+        always_print_fields_with_no_presence: Optional[bool],
+        including_default_value_fields: Optional[bool],
+    ) -> bool:
+        """
+        Return true if fields with no presense should be included in the results.
```
[the remark this answers is not in the document]
Comment (parthea) [reply proto/message.py]: Fixed in https://github.com/googleapis/proto-plus-python/pull/433/commits/2da20df33ac5621f72769345f7187337c8598745

File: proto/message.py [hunk 18]
```diff
@@ -398,19 +492,45 @@ def to_json(
...
+                preserving_proto_field_name=preserving_proto_field_name,
+                sort_keys=sort_keys,
+                indent=indent,
+                float_precision=float_precision,
+            )
+        else:
+            # The `including_default_value_fields` argument was removed from protobuf 5.x
+            # and replaced with `always_print_fields_with_no_presence` which very similar but has
+            # consistent handling of optional fields by not affecting them.
+            # The old flag accidentally had inconsistent behavior between proto2 optional and proto3 optional fields.
```
[the remark this answers is not in the document]
Comment (parthea) [reply proto/message.py]: Fixed in https://github.com/googleapis/proto-plus-python/pull/433/commits/d525973cae84f910808ae515b4729852132c9643

File: proto/message.py [hunk 19]
```diff
@@ -434,39 +554,65 @@ def to_dict(
...
+        if PROTOBUF_VERSION[0] in ("3", "4"):
+            return MessageToDict(
+                cls.pb(instance),
+                including_default_value_fields=print_fields,
+                preserving_proto_field_name=preserving_proto_field_name,
+                use_integers_for_enums=use_integers_for_enums,
+                float_precision=float_precision,
+            )
+        else:
+            # The `including_default_value_fields` argument was removed from protobuf 5.x
```
[the remark this answers is not in the document]
Comment (parthea) [reply proto/message.py:605]: Fixed in https://github.com/googleapis/proto-plus-python/pull/433/commits/d525973cae84f910808ae515b4729852132c9643

File: testing/constraints-3.13.txt
[the remark this answers is not in the document]
Comment (parthea) [reply testing/constraints-3.13.txt:1]: The empty file is correct. When the file is empty it means that we aren't pinning dependencies. We typically pin dependencies in `constraints-3.7.txt`. The reason for pinning the dependencies is to make sure that we run tests against the minimum supported versions. This is captured in https://github.com/googleapis/proto-plus-python/blob/0833e4cd0309d3bec2961935d14760c56858be3e/testing/constraints-3.7.txt#L1-L5

File: tests/test_json.py (hunk 9 above)
Comment (parthea) [reply tests/test_json.py]: Fixed in https://github.com/googleapis/proto-plus-python/pull/433/commits/160cf40b1056fff269384859c99a1a27b07df409

File: tests/test_message.py (hunk 10 above)
Comment (parthea) [reply tests/test_message.py]: I fixed the test case in https://github.com/googleapis/proto-plus-python/pull/433/commits/7af40e3357583448ca3167b97a09668daa9c9e35. I wasn't able to get the value of `UNKNOWN` to show up as the default value of the repeated field `chromatophores` is `[]`

File: tests/test_json.py (hunk 11 above)
Comment (parthea) [reply tests/test_json.py:134]: I believe the `including_default_value_fields=False` and `always_print_fields_with_no_presence=False` test cases are complete. I added more test cases for `including_default_value_fields=True` and `always_print_fields_with_no_presence=True`

https://github.com/googleapis/proto-plus-python/pull/433/commits/78628035e7789eb12f61cf31b568dc8dc1e2f57f

Comment (vchudnov-g) [reply tests/test_json.py:134]: OK, looks good. Thanks for adding these test cases!

File: tests/test_json.py [hunk 20]
```diff
@@ -121,10 +121,77 @@ class Squid(proto.Message):
...
+        Squid.to_json(
+            s,
+            always_print_fields_with_no_presence=True,
+        )
+        .replace(" ", "")
+        .replace("\n", "")
+    )
+    assert json2 == '{"name":"Steve","massKg":0}'
+
+    json2 = Squid.to_json(s).replace(" ", "").replace("\n", "")
```
Comment (vchudnov-g) [inline tests/test_json.py:193]: +1 to this test case. It just occurred to me as well to test what happens when we exclude both parameters.

File: tests/test_message.py [hunk 21]
```diff
@@ -267,6 +267,60 @@ class Color(proto.Enum):
...
+    assert s_dict_2 == expected_dict
+
+    s_dict_2 = Squid.to_dict(
+        s_new_2,
+        always_print_fields_with_no_presence=True,
+    )
+    expected_dict = {"mass_kg": 20, "chromatophores": []}
+    assert s_dict_2 == expected_dict
+
+    s_dict_2 = Squid.to_dict(
```
Comment (vchudnov-g) [inline tests/test_message.py:296]: Can we add a test case here where we don't pass either parameter?

File: tests/test_message.py [hunk 22]
```diff
@@ -267,6 +267,60 @@ class Color(proto.Enum):
...
+        match="Arguments.*always_print_fields_with_no_presence.*including_default_value_fields.*must match",
+    ):
+        s_dict_2 = Squid.to_dict(
+            s_new_2,
+            including_default_value_fields=False,
+            always_print_fields_with_no_presence=True,
+        )
+
     new_s = Squid(s_dict)
     assert new_s == s
```
Comment (vchudnov-g) [inline tests/test_message.py:325]: The two lines below are included further up in this test function, so they're needlessly repetitive here.

```suggestion
```

**Review by vchudnov-g — APPROVED** [review]
Looks good. Left two minor comments, but I think we're good to go.,

File: tests/test_message.py (hunk 21 above)
Comment (parthea) [reply tests/test_message.py:296]: Done in https://github.com/googleapis/proto-plus-python/pull/433/commits/9d85d8bf84b861333385ac30d7512a6d0d211309
